### PR TITLE
Refactor receipt form

### PR DESCRIPTION
**Pull request by viennguyen24**
**Issue:** #39 

**Branch:** refactor/receipt-form

**PR:** #62

**Data:** #47 

**Files changed:** 
1. client/src/pages/Receipt.tsx, client/src/pages/index.ts, 
2. client/src/App.tsx
3. client/src/components/ingredients/ReceiptForm.tsx, 
4. client/src/components/ingredients/ReceiptField.tsx, 
5. client/src/components/ingredients/MockReceipt.tsx

**Library:** chakra-react-select, use-debounce

**Overview**
The receipt form is a front-end feature to help users manually edit their receipts after scanning it with the receipt scanner. This PR resolves the previous receipt form, which is bulky, has lots of nested and unorganized CSS and UI props, and are not reusable because it is hard fixed (old PR for reference before delete #49 ). The form is refactored into:

**1. Receipt page:** You will initialize the schema object, useForm, useField arrays and all methods to handle submit and changes fields information here
**2. ReceiptForm:** This component receives a control to a useForm, the fields and their methods, and render a children component to handle each field
**3. ReceiptField:** This is the children component to handle a single field. The children will render information of a single field and can access the fields array to make changes

**Flow for reuse:**
1. Define new useForm, useField array and each update method
2. Define a different children component like ReceiptField to handle how a single field is rendered 
3. Pass the form data, fields, and the children component as props to the ReceiptForm with based on type `ReceiptFormProps`

**Demo:**
[Demo receipt form](https://www.loom.com/share/411c59d29abf4b448cc80dbacef2bacc)

**Sample data**
```typescript
{
  "receipts": [
    {
      "name": "pumpkin vanilla",
      "quantity": "1",
      "price": "165.00",
      "suggested": {
        "display": false,
        "items": [
          "vanilla",
          "sugar pumpkin",
          "vanilla frosting",
          "vanilla beans",
          "vanilla sugar",
          "vanilla yogurt",
          "vanilla icecream",
          "tahitian vanilla bean",
          "bourbon vanilla extract",
          "canned pumpkin puree",
          "vanilla froyo",
          "vanilla cake mix",
          "Pumpkin Kernels",
          "vanilla powder",
          "vanilla bean paste",
          "gluten-free vanilla cake mix",
          "vanilla wafers",
          "fat-free vanilla yogurt",
          "vanilla light ice cream",
          "imitation vanilla"
        ]
      }
    },
    {
      "name": "tomato & cheese",
      "quantity": "1",
      "price": "185.00",
      "suggested": {
        "display": false,
        "items": [
          "mozzarella cheese",
          "tomatoes",
          "cheese",
          "gorgonzola cheese",
          "parmesan cheese",
          "tomato & basil sauce",
          "cheese soup",
          "gruyere cheese",
          "camembert cheese",
          "cheese pumpkin",
          "burrata cheese",
          "stilton cheese",
          "provolone cheese",
          "fresh mozzarella cheese",
          "cheese raviolis",
          "nacho cheese sauce",
          "gouda cheese",
          "fontina cheese",
          "brie cheese",
          "havarti cheese"
        ]
      }
    },
    {
      "name": "chorizo gambas",
      "quantity": "1",
      "price": "375.00",
      "suggested": {
        "display": false,
        "items": [
          "chorizo sausage",
          "dried chorizo",
          "pancetta",
          "pork escalope",
          "jamon iberico",
          "andouille sausages",
          "pork sausages",
          "chipolata sausage",
          "bulk chorizo",
          "jamon",
          "boar stew meat",
          "tagliatelle",
          "pork cutlets",
          "pork broth",
          "pipian",
          "cooked tagliatelle",
          "sausages",
          "italian turkey sausages",
          "cooked papardelle",
          "jalapeno sauce"
        ]
      }
    },
    ]
 }
```

### CODE REVIEW

**Review by hungngodev — CHANGES_REQUESTED** [review]
Merge the conflict please!

**Comment by hoatuyet423** [comment]
- Love the new UI!!!
- The pencil icon for AI suggestion is not intuitive, change it to something more uinversal (like a star or a box that said suggestion or you can also just keep the suggestion box you have right now and have it show without clicking the pencil icon)
- NIT: Add a line or a very light border to seperate each ingredients will make it look better.

**Review by hungngodev — CHANGES_REQUESTED** [review]
Please update the latest UI review comment today!

File: client/src/components/ingredients/ReceiptForm.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      <CardFooter>
+        <Button
+          colorScheme="teal"
+          letterSpacing="-0.005em"
+          fontWeight="semibold"
+          onClick={() => {
+            append();
+          }}
+        >
+          Add new receipt
```
Comment (hungngodev) [inline client/src/components/ingredients/ReceiptForm.tsx]: Change this to add new input

File: client/src/components/ingredients/ReceiptField.tsx [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+  control,
+  index,
+  remove,
+  update,
+  fields,
+  field,
+  watch,
+}: ReceiptFieldProps<T>) {
+  const updateReceipt = useDebouncedCallback((val: any) => {
+    console.log(val);
```
Comment (hungngodev) [inline client/src/components/ingredients/ReceiptField.tsx]: Please make a mock API call to backend to help NA know. 
For example you could just write a line like 
const newData = await customFetch.get(something)
Do some update logic with the form to reflect it like modify the Options variable below or modify the field.suggested.item something.

**Review by hungngodev — CHANGES_REQUESTED** [review]

**Pull request closed without merging**